File: Local-TFG/Criptografia/prueba-1/python/main.py
import time
from pyscript import PyWorker, display
import js  # type: ignore
import json

worker = None


async def initialize_worker():
    global worker
    if worker is None:
        worker = PyWorker(
            "./python/worker.py",
            type="pyodide",
            config="./json/pyscript-worker.json"
        )
        await worker.ready


async def launch_worker(event):
    global worker, worker_time
    js.clearCell('pyscript-output')
    js.startPyTimer()

    start_w = time.perf_counter()
    await initialize_worker()
    worker_time = (time.perf_counter() - start_w) * 1000

    result_json = await worker.sync.do_analisis()
    result = json.loads(result_json)

    display_result(result)


def display_result(result):
    global worker_time
    display(
        f"Worker init time: {worker_time:.2f} ms", target="pyscript-output")
    display(
        f"Simulated file: {result['file_size_mb']:.2f} MB", target="pyscript-output")
    display(
        f"Repetitions: {result['repetitions']}", target="pyscript-output")
    display(f"Avg hash time: {result['hash_avg_time_ms']:.2f} ms",
            target="pyscript-output")
    display(f"Avg verify time: {result['verify_avg_time_ms']:.2f} ms",
            target="pyscript-output")
    display(
        f"Last hash (shortened): {result['last_digest'][:8]}...", target="pyscript-output")

    display(f"Total hash + verify time: {result['total_time_ms']:.2f} ms",
            target="pyscript-output")
    display(
        f"Total ET: {result['total_time']:.2f} ms", target="pyscript-output")

    js.stopPyTimer()

File: Local-TFG/Criptografia/prueba-1/javascript/main.js
let worker = null;
let jsTimer = null;
let jsStartTime = 0;

function startJsTimer() {
    jsStartTime = performance.now();
    const timerElement = document.getElementById("javascript-timer-display");
    function updateTimer() {
        if (!jsTimer) return;
        const elapsed = ((performance.now() - jsStartTime) / 1000).toFixed(3);
        timerElement.textContent = `JS Timer: ${elapsed} s`;
        jsTimer = requestAnimationFrame(updateTimer);
    }
    cancelAnimationFrame(jsTimer);
    jsTimer = requestAnimationFrame(updateTimer);
}

function stopJsTimer() {
    cancelAnimationFrame(jsTimer);
    jsTimer = null;
    const elapsed = ((performance.now() - jsStartTime) / 1000).toFixed(3);
    document.getElementById(
        "javascript-timer-display"
    ).textContent = `JS Timer: ${elapsed} s`;
}

function initializeWorker() {
    if (!worker) {
        worker = new Worker(new URL("worker.js", import.meta.url), {
            type: "module",
        });
    }
}

async function runJsBenchmark() {
    try {
        window.clearCell("javascript-output");
        startJsTimer();

        const startWorkerTime = performance.now();
        initializeWorker();
        const workerTime = performance.now() - startWorkerTime;

        const repetitions = parseInt(
            document.getElementById("num-repetitions-javascript").value,
            10
        );
        const fileSizeMb = parseFloat(
            document.getElementById("file-size-javascript").value
        );
        const id = `js-${Date.now()}`;

        const resultJson = await new Promise((resolve, reject) => {
            function onMessage(e) {
                if (e.data.id !== id) return;
                worker.removeEventListener("message", onMessage);
                if (e.data.error) {
                    reject(new Error(e.data.error));
                } else if (e.data.json !== undefined) {
                    resolve(e.data.json);
                } else {
                    reject(new Error("Worker dont return results"));
                }
            }
            worker.addEventListener("message", onMessage);
            worker.postMessage({
                id,
                type: "do_analisis",
                repetitions,
                fileSizeMb,
            });
        });

        const result = JSON.parse(resultJson);
        displayResult(result, workerTime);
    } catch (err) {
        console.error("Benchmark error:", err);
    } finally {
        stopJsTimer();
    }
}

function createDiv() {
    const div = document.createElement("div");
    return div;
}

function displayResult(r, workerTime) {
    const output = document.getElementById("javascript-output");

    const workerDiv = createDiv();
    workerDiv.textContent = `Worker init time: ${workerTime.toFixed(2)} ms`;
    output.appendChild(workerDiv);

    const simulatedDiv = createDiv();
    simulatedDiv.textContent = `Simulated file: ${r.file_size_mb.toFixed(2)} ms`;
    output.appendChild(simulatedDiv);

    const repetitionsDiv = createDiv();
    repetitionsDiv.textContent = `Repetitions: ${r.repetitions.toFixed(2)} ms`;
    output.appendChild(repetitionsDiv);

    const hashDiv = createDiv();
    hashDiv.textContent = `Avg hash time: ${r.hash_avg_time_ms.toFixed(2)} ms`;
    output.appendChild(hashDiv);

    const verifyDiv = createDiv();
    verifyDiv.textContent = `Avg verify time: ${r.verify_avg_time_ms.toFixed(2)} ms`;
    output.appendChild(verifyDiv);

    const lastDigestDiv = createDiv();
    lastDigestDiv.textContent = `Last hash (shortened): ${r.last_digest.slice(0, 8)}`;
    output.appendChild(lastDigestDiv);

    const totalOpDiv = createDiv();
    totalOpDiv.textContent = `Total hash + verify time: ${r.total_time_ms.toFixed(2)} ms`;
    output.appendChild(totalOpDiv);

    const totalDiv = createDiv();
    totalDiv.textContent = `Total ET: ${r.total_time.toFixed(2)} ms`;
    output.appendChild(totalDiv);
}

window.runJsBenchmark = runJsBenchmark;
